lib/active_model/serializer/adapter/json_api.rb

```diff
@@ -113,12 +113,13 @@ def serializable_hash_for_single_resource
         end
 
         def resource_identifier_type_for(serializer)
-          return serializer._type if serializer._type
-          if ActiveModelSerializers.config.jsonapi_resource_type == :singular
-            serializer.object.class.model_name.singular
-          else
-            serializer.object.class.model_name.plural
+          value_or_block = serializer._type
+          if value_or_block
+            return serializer.instance_eval(&value_or_block) if value_or_block.respond_to?(:call)
+            return value_or_block
           end
+          inflection = ActiveModelSerializers.config.jsonapi_resource_type
+          serializer.object.class.model_name.public_send(inflection)
```

Comment:
Concerning your first point, I based my change on the `link` method and how it is used in the adapter (https://github.com/groyoh/active_model_serializers/blob/dynamic_type/lib/active_model/serializer/adapter/json_api.rb#L214). I'd be happy to change it so that the setter creates a callable out of the value if you consider that's a win.
Concerning the changes I made in this function they were mainly DRYing up the code, but I'll revert it to keep only the necessary code (and maybe create another PR for the unnecessary changes).
Concerning the changes in the specs, I mainly refactored the code to use a helper method to avoid repeating the code everywhere in the test file and also added specs to test that using the method in a subclass doesn't break it in the superclass.